[lldb] fix dead lock in TypeCategoryMap.cpp

v-bulle opened the pull request:
FormatManager::GetCategoryForLanguage and
FormatManager::GetCategory(can_create = true) can be called concurrently and they both take the TypeCategory::m_map_mutex and the FormatManager::m_language_categories_mutex but in reverse order.

On one thread, GetCategoryForLanguage takes m_language_categories_mutex and then ends calling TypeCategoryMap::Get which takes m_map_mutex

On another thread GetCategory calls TypeCategoryMap::Add which takes m_map_mutex and then calls FormatManager::Changed() which takes m_language_categories_mutex

If both threads are running concurrently, we have a dead lock.

The patch releases the m_map_mutex before calling Changed which avoids the dead lock.

Commit:
[lldb] fix dead lock in TypeCategoryMap.cpp

FormatManager::GetCategoryForLanguage and
FormatManager::GetCategory(can_create = true) can be called concurently
and they both take the TypeCategory::m_map_mutex and the
FormatManager::m_language_categories_mutex but in reverse order.

Without the patch, we can have a dead lock.

JDevlieghere commented:
The change itself looks fine, but I'm not sure I follow the motivation. I see where `FormatManager::GetCategoryForLanguage` locks `m_language_categories_mutex` and I see where  `FormatManager::GetCategory` calls `TypeCategoryMap::Get` and locks `m_map_mutex`, but I don't immediately see where they're acquiring the other mutex. Can you elaborate on the situation that causes the deadlock?

bulbazord commented:
I'm struggling to understand how your description matches the change as well. You mention `FormatManager::GetCategoryForLanguage` and `FormatManager::GetCategory` but your change doesn't touch those. Instead you modify `TypeCategoryMap::Add`, `TypeCategoryMap::Delete`, and `TypeCategoryMap::Clear`. It makes sense that those only hold onto mutex for the short time that they need them, but why does changing those prevent the deadlock?

Something that might help is a way to reproduce the deadlock. I know deadlocks are not deterministically reproducible in most cases, but if you could walk us through the scenario, it might be easier to see.

v-bulle commented:
In one thread we have:
    frame #3: 0x0000560bc08118c3 lldb`lldb_private::TypeCategoryMap::Get(lldb_private::ConstString, std::__u::shared_ptr<lldb_private::TypeCategoryImpl>&) + 35
    frame #4: 0x0000560bc0800a1e lldb`lldb_private::FormatManager::GetCategory(lldb_private::ConstString, bool) + 62
    frame #5: 0x0000560bc07fc58e lldb`lldb_private::DataVisualization::Categories::GetCategory(lldb_private::ConstString, std::__u::shared_ptr<lldb_private::TypeCategoryImpl>&, bool) + 46
    frame #6: 0x0000560bc099d5b9 lldb`___lldb_unnamed_symbol166675 + 57
    frame #7: 0x0000560bc5c42c15 lldb`std::__u::__call_once(unsigned long volatile&, void*, void (*)(void*)) + 149
    frame #8: 0x0000560bc099c6ac lldb`lldb_private::ObjCLanguage::GetFormatters() + 76
    frame #9: 0x0000560bc0805daf lldb`lldb_private::LanguageCategory::LanguageCategory(lldb::LanguageType) + 143
    frame #10: 0x0000560bc080139e lldb`lldb_private::FormatManager::GetCategoryForLanguage(lldb::LanguageType) + 126
And in the other:
    frame #3: 0x0000560bc07fe586 lldb`lldb_private::FormatManager::Changed() + 38
    frame #4: 0x0000560bc08114cb lldb`lldb_private::TypeCategoryMap::Add(lldb_private::ConstString, std::__u::shared_ptr<lldb_private::TypeCategoryImpl> const&) + 139
    frame #5: 0x0000560bc0800aba lldb`lldb_private::FormatManager::GetCategory(lldb_private::ConstString, bool) + 218

For the first thread, GetCategoryForLanguage takes m_language_categories_mutex and then ends calling TypeCategoryMap::Get which takes m_map_mutex

In the other thread GetCategory calls TypeCategoryMap::Add which takes m_map_mutex and then calls FormatManager::Changed() which takes m_language_categories_mutex

With the change, in TypeCategoryMap::Add, we release m_map_mutex before calling Changed (which doesn't need it) and avoid the dead lock. The pull request also releases the lock in Delete and Clear to be consistent.

JDevlieghere commented:
Alright, so this happens when `listener` in `TypeCategoryMap` is the `TypeCategory`. Please update the commit message and add a comment before the call to `Changed` to say why we're doing it outside the scope of the lock. With those changes this should be good to go.

Commit:
Adds comments

JDevlieghere commented on lldb/source/DataFormatters/TypeCategoryMap.cpp:
```diff
@@ -25,19 +25,25 @@ TypeCategoryMap::TypeCategoryMap(IFormatChangeListener *lst)
+    std::lock_guard<std::recursive_mutex> guard(m_map_mutex);
+    m_map[name] = entry;
+  }
+  // The lock is now released for the eventual call to Changed.
```
This comment explains _what_ you're doing, which the code already tells me, but not _why_.

I'd expect something like this, although it doesn't need to be quite so verbose:

```
// Release the mutex to avoid a potential deadlock between TypeCategoryMap::m_map_mutex and FormatManager::m_language_categories_mutex which can be acquired in reverse order when calling FormatManager::Changed.
```

v-bulle replied:
done

Commit:
Modify comment

JDevlieghere reviewed (approved):
LGTM
